Base/usr/share/man/man2/futex.md

```diff
@@ -0,0 +1,191 @@
+  implementation for details.
+
+Additionally, the `FUTEX_PRIVATE_FLAG` flag can be *or*'ed in with one of the
+*operation* values listed above. This flag restricts the call to only work on
+other threads of the same process (as opposed to any threads in the system that
+may have the same memory page mapped into their address space, possibly at a
+different address), which enables additional optimizations in the syscall
+implementation. The inverse of this flag is exposed as the `process_shared`
+argument in `futex_wait()` and `futex_wake()` wrapper functions.
+
+## Return value
+
+* `FUTEX_WAKE`, `FUTEX_WAKE_BITSET`, `FUTEX_WAKE_OP`: the number of the waiting
+  threads that have been woken up, which may be 0 or a positive number.
+* `FUTEX_WAIT`, `FUTEX_WAIT_BITSET`: 0 if blocked and got woken up by an
+  explicit wake call or woke up spuriously, an error otherwise.
+* `FUTEX_REQUEUE`, `FUTEX_CMP_REQUEUE`: the total number of threads woken up
+  and requeued.
+
+## Errors
+
+* `EAGAIN`: for wait operations, did not begin waiting, because the futex value
+   has already been changed.
+* `ETIMEDOUT`: for wait operations with a timeout, timed out.
+* `EFAULT`: the specified futex address is invalid.
+* `ENOSYS`: `FUTEX_CLOCK_REALTIME` was specified, but the operation is not
+  `FUTEX_WAIT` or `FUTEX_WAIT_BITSET`.
+* `EINVAL`: The arithmetic-logical operation for `FUTEX_WAIT_OP` is invalid.
+
+## Examples
+
+The following program demonstrates how futexes can be used to implement a
+simple "event" synchronization primitive. An event has a boolean state: it can
+be *set* or *unset*; the initial state being unset. The two operations on an
+event are *waiting* until it is set, and *setting* it (which wakes up any
+threads that were waiting for the event to get set).
+
+Such a synchronization primitive could be used, for example, to notify threads
+that are waiting for another thread to perform some sort of complex
+initialization.
+
+The implementation features two fast paths: both setting an event that no
+thread is waiting on, and trying to wait on an event that has already been set,
+are performed entirely in userspace without calling into the kernel. For this
+to work, the value of the futex integer is used to track both the state of the
+event (whether it has been set) and whether any threads are waiting on it.
+
+```c++
+#include <AK/Atomic.h>
+#include <serenity.h>
+
+class Event {
+private:
+    enum State : u32 {
+        UnsetNoWaiters,
+        UnsetWithWaiters,
+        Set,
+    };
+
+    AK::Atomic<State> m_state { UnsetNoWaiters };
+
+public:
+    void set()
+    {
+        State previous_state = m_state.exchange(Set, AK::memory_order_release);
+        // If there was anyone waiting, wake them all up.
+        // Fast path: no one was waiting, so we're done.
+        if (previous_state == UnsetWithWaiters)
+            futex_wake((u32*)m_state.ptr(), UINT32_MAX, false);
```

Comment:
The example should follow our guidelines and use a non-C cast here (I think static_cast is possible?)

Reply:
I'm afraid `reinterpret_cast<u32*>(const_cast<State*>(m_state.ptr()))` is the only thing that compiles, and that is _a mouthful_.

Guess a I could introduce a tiny helper:

```c++
private:
    u32* state_futex_ptr() { return reinterpret_cast<u32*>(const_cast<State*>(m_state.ptr())); }
```

Reply:
Done that, so tapping "resolve conversation", but please unresolve if that's not good enough.